src/Databases/DatabaseFilesystem.cpp

```diff
@@ -0,0 +1,153 @@
+#include <Databases/DatabaseFilesystem.h>
+
+#include <IO/Operators.h>
+#include <IO/WriteBufferFromString.h>
+#include <Interpreters/Context.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ParserCreateQuery.h>
+#include <Parsers/parseQuery.h>
+#include <Storages/IStorage.h>
+#include <TableFunctions/TableFunctionFactory.h>
+
+#include <filesystem>
+
+namespace fs = std::filesystem;
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int UNKNOWN_TABLE;
+}
+
+DatabaseFilesystem::DatabaseFilesystem(const String & name_, const String & path_, ContextPtr context_)
+    : IDatabase(name_), WithContext(context_->getGlobalContext()), path(path_), log(&Poco::Logger::get("DatabaseFileSystem(" + name_ + ")"))
+{
+    path = fs::path(path).lexically_normal().string();
+}
+
+std::string DatabaseFilesystem::getTablePath(const std::string & table_name) const
+{
+    return fs::path(path) / table_name;
+}
+
+void DatabaseFilesystem::addTable(const std::string & table_name, StoragePtr table_storage) const
+{
+    std::lock_guard lock(mutex);
+    loaded_tables.emplace(table_name, table_storage);
+}
+
+bool DatabaseFilesystem::isTableExist(const String & name, ContextPtr) const
+{
+    {
+        std::lock_guard lock(mutex);
+        if (loaded_tables.find(name) != loaded_tables.end())
+            return true;
+    }
+
+    fs::path table_file_path(getTablePath(name));
+    return fs::exists(table_file_path) && fs::is_regular_file(table_file_path);
+}
+
+StoragePtr DatabaseFilesystem::getTableImpl(const String & name, ContextPtr context_) const
+{
+    // Check if the table exists in the loaded tables map
+    {
+        std::lock_guard lock(mutex);
+        auto it = loaded_tables.find(name);
+        if (it != loaded_tables.end())
+            return it->second;
+    }
+
+    auto table_path = getTablePath(name);
+
+    // If the file exists, create a new table using TableFunctionFile and return it.
+    auto args = makeASTFunction("file", std::make_shared<ASTLiteral>(table_path));
```

Comment:
we should check here if `table_path` actually exists, and if it does not exist, this method should return nullptr.
Because now this method is like `getOrCreateTable` while its name suggests that it should do only `getTable`

Reply:
yes, it does not create a new file, but you created a StoragePtr in tryGetTable even though the file could not exist, which means that the table as well should not exist, while method getTableImpl should return nullptr if table does not exist.